test/tests_for_sdkmr/EE2Status_test.py

```diff
@@ -0,0 +1,125 @@
+    mongo = create_autospec(MongoUtil, spec_set=True, instance=True)
+    kafka = create_autospec(KafkaClient, spec_set=True, instance=True)
+    catutil = create_autospec(CatalogUtils, spec_set=True, instance=True)
+    catalog = create_autospec(Catalog, spec_set=True, instance=True)
+    condor = create_autospec(Condor, spec_set=True, instance=True)
+    sdkmr.get_mongo_util.return_value = mongo
+    sdkmr.get_logger.return_value = logger
+    sdkmr.get_kafka_client.return_value = kafka
+    sdkmr.get_condor.return_value = condor
+    sdkmr.get_catalog_utils.return_value = catutil
+    catutil.get_catalog.return_value = catalog
+
+    # set up return values for mocks. Ordered as per order of operations in code
+    job1 = _finish_job_complete_minimal_get_test_job(
+        job_id, sched, app_id, gitcommit, user
+    )
+    job2 = _finish_job_complete_minimal_get_test_job(
+        job_id, sched, app_id, gitcommit, user
+    )
+    job2.status = Status.completed.value
+
+    sdkmr.get_job_with_permission.side_effect = [job1, job2]
+    mongo.get_job.return_value = job2  # gets the job 3x...?
+    condor.get_job_resource_info.return_value = resources
+
+    # call the method
+    JobsStatus(sdkmr).finish_job(job_id, job_output=job_output)  # no return
+
+    # check mocks called as expected. Ordered as per order of operations in code
+
+    sdkmr.get_job_with_permission.assert_has_calls(
+        [
+            call(
+                job_id=job_id, requested_job_perm=JobPermissions.WRITE, as_admin=False
+            ),
+            call(
+                job_id=job_id, requested_job_perm=JobPermissions.WRITE, as_admin=False
+            ),
+        ]
+    )
+    logger.debug.assert_has_calls(
+        [
+            call("Finishing job with a success"),
+            # depending on stable dict ordering for this test to pass
+            call(f"Extracted the following condor job ads {resources}"),
+        ]
+    )
+    mongo.finish_job_with_success.assert_called_once_with(job_id, job_output)
+    kafka.send_kafka_message.assert_called_once_with(
+        KafkaFinishJob(
+            job_id=job_id,
+            new_status=Status.completed.value,
+            previous_status=Status.running.value,
+            scheduler_id=sched,
+            error_code=None,
+            error_message=None,
+        )
+    )
+    mongo.get_job.assert_called_once_with(job_id)
+    catalog.log_exec_stats.assert_called_once_with(
+        {
+            "user_id": user,
+            "app_module_name": "module",
+            "app_id": app_id,
+            "func_module_name": "module",
+            "func_name": "method_id",
+            "git_commit_hash": gitcommit,
+            "creation_time": 1615246649.0,  # from Job ObjectId
+            "exec_start_time": 123.0,
+            "finish_time": 456.5,
+            "is_error": 0,
+            "job_id": job_id,
+        }
+    )
+    mongo.update_job_resources.assert_called_once_with(job_id, resources)
```

Comment:
What about some negative tests? Or tests where failures are expected with bad inputs?

Reply:
The point of this PR was not to add tests, per se - it was to fix the bug re accessing the catalog client by an instance variable that doesn't exist, and write tests that cover that bug so that it can't reoccur. The happy path test was enough to cover the bug, so failure tests weren't needed.

Reply:
Sure, but that's true for lots of places in the code. The point of this PR is to fix and test a specific bug, not to add tests in general.

Reply:
The point is to fix a bug and, hopefully, prevent others from taking its place due to poorly tested code. Please write some more tests to capture other failure conditions.

Reply:
> and, hopefully, prevent others from taking its place due to poorly tested code

Hmm, looks like we're at an impasse, because AFAIC, that is not part of the point of this PR. I'll bug @sychan to be a tiebreaker here.

Reply:
Works for me. The past few times you complained of errors happening from untested code and from things that reviewers didn't catch, I figured this would be an opportunity to avoid that. Letting @sychan decide whether more work is warranted here is fine.

Reply:
> The past few times you complained of errors happening from untested code and from things that reviewers [note: and I, and 5 linters -g] didn't catch

That's still a valid complaint, and I'll continue to address it by adding tests for code that I touch. It's just that
1) if I add unit tests for every piece of code I see that doesn't have unit tests I'll never work on the actual feature I'm supposed to be working on, and
2) I want to keep this PR focused on the bugfix.

It became clear to me early on that unless I restrict adding tests to code I actually make changes to I won't finish, as demonstrated by the fact that I'm already several weeks into this process and just barely started on the feature.

Reply:
Bill is correct on the need for testing failed runs, and given how important EE2 to the core functioning of our platform - we do, desperately, need more pervasive testing on the EE2 codebase. And it is ironic that Gavin is the one arguing for not adding tests, the shoe isn't so comfortable when its on the other foot!

The mitigating circumstance is that we are weeks (if not coming up months behind) on getting basic features implemented and we're 3 weeks from the end of the quarter, after which we are likely to take a huge hit on resourcing for the Bulk Import work, and I would actually like Gavin to go somewhat wide on the EE2 codebase, instead of deep at this point. I'm going to say we let this slide, and have Gavin move on - with the understanding that Gavin will work on incrementally improving the code coverage in EE2 as he does future PRs.